### PR TITLE
Add support for empty Vecs - addresses #529

### PR DESCRIPTION
**Pull request by ucbjrl**

### CODE REVIEW

**Comment by aswaterman** [comment]
Great, we need this.  I think there are a few more things to check:

- Right now it looks like this would error if you try to emit Firrtl, because ```toType``` will dereference null.  Presumably ```Wire(Vec(Seq()))``` would trigger this.  Not sure what the right fix is.
- It looks like ```cloneType``` needs to be special-cased to avoid dereferencing null in ```gen.cloneType```.
- ```apply(UInt)``` will also dereference null; it should probably throw an IndexOutOfBounds exception instead, as ```apply(0)``` does in this case.  (Actually, returning ```apply(0)``` in this case is a good way to do that.)
- ```typeEquivalent``` needs to be special-cased to return true if both lengths are 0, or else the comparisons on ```sample_element``` will dereference null.
- ```allElements``` should return an empty list if length is 0 (```sample_elements``` is again the problem here).  Maybe this (and the previous one) are best handled by making ```sample_element``` an Option instead.

File: chiselFrontend/src/main/scala/chisel3/core/Aggregate.scala [hunk 1]
```diff
@@ -66,29 +66,32 @@ object Vec {
...
-    def doConnect(sink: T, source: T) = {
-      // TODO: this looks bad, and should feel bad. Replace with a better abstraction.
-      // NOTE: Must use elts.head instead of vec.sample_element because vec.sample_element has
-      //       WireBinding which does not have a direction
-      val hasDirectioned = elts.head match {
-        case t: Aggregate => t.flatten.exists(_.dir != Direction.Unspecified)
-        case t: Element => t.dir != Direction.Unspecified
+    if (elts.isEmpty) {
+      // Support empty Vecs.
+      new Vec(null.asInstanceOf[T], 0)
```
Comment (ducky64) [inline chiselFrontend/src/main/scala/chisel3/core/Aggregate.scala]: null.asInstanceOf[T] raises serious red flags.

Possible alternative solutions:
- Make sample_element an Option[T]
- Remove sample_element entirely, using instead head of Vec elements, and handling the empty case properly

Comment (ducky64) [inline chiselFrontend/src/main/scala/chisel3/core/Aggregate.scala]: Related: making this an Option will probably address some of @aswaterman 's cases at compile time. Using Vec elements head is also an option as long as we consistently check all accesses for empty. sample_element can also return elements head as an Option, which might be cleaner.

**Comment by ucbjrl** [comment]
Could this be solved by converting `Vec(Seq[T]())` into `new Vec(gen[T], 0)` where `gen[T]` constructs an instance of `T`? I realize this is problematic where parameterized types are concerned.

Failing that, can we squirrel away the type of `T` so we can still use it in those cases where we were relying on the `sample_element` to provide `T`?

We still need to be able to deal with optional `sample_element`s or empty `elements`.

**Comment by edwardcwang** [comment]
Related to #946

**Comment by ucbjrl** [comment]
Closed since it seems there is another solution has crept in and #529 has been closed for almost two years.

**Pull request closed without merging**